File: src/main/java/actions/MovementHandler.java
package actions;

import processing.core.PApplet;
import ux.PatternDrawer;

import java.util.Set;

// this class exists merely to encapsulate movement handling as it is a bit complex
// and i don't want to clutter the main class anymore than it already
public class MovementHandler {

    private final static Set<Integer> pressedKeys = KeyHandler.getPressedKeys();

    private int lastDirection = 0;
    private long lastIncreaseTime;
    private final float initialMoveAmount = 1;
    private float moveAmount = initialMoveAmount;

    private final PatternDrawer drawer;

    public MovementHandler(PatternDrawer drawer) {
        this.drawer = drawer;
    }


    public void handleRequestedMovement() {
        for (Integer key : pressedKeys) {
            switch (key) {
                case PApplet.LEFT, PApplet.RIGHT, PApplet.UP, PApplet.DOWN -> handleMovementKeys();
                default ->
                    // Ignore other keys and set lastDirection to 0
                        lastDirection = 0;
            }
        }
    }


    private void handleMovementKeys() {

        float moveX = 0;
        float moveY = 0;

        int[][] directions = {
                {PApplet.LEFT, 0, -1},
                {PApplet.UP, -1, 0},
                {PApplet.RIGHT, 0, 1},
                {PApplet.DOWN, 1, 0},
                {PApplet.UP + PApplet.LEFT, -1, -1},
                {PApplet.UP + PApplet.RIGHT, -1, 1},
                {PApplet.DOWN + PApplet.LEFT, 1, -1},
                {PApplet.DOWN + PApplet.RIGHT, 1, 1}
        };

        for (int[] direction : directions) {
            boolean isMoving = pressedKeys.contains(direction[0]);
            if (isMoving) {
                moveX += direction[2] * moveAmount;
                moveY += direction[1] * moveAmount;
            }
        }

        // Check if the direction has changed
        int currentDirection = 0;
        for (int[] direction : directions) {
            boolean isMoving = pressedKeys.contains(direction[0]);
            if (isMoving) {
                currentDirection += direction[0];
            }
        }

        if (currentDirection != lastDirection) {
            // Reset moveAmount if direction has changed
            moveAmount = initialMoveAmount;
            lastIncreaseTime = System.currentTimeMillis();
        } else {
            // Increase moveAmount if enough time has passed since last increase
            long currentTime = System.currentTimeMillis();
            long increaseInterval = 100;
            if (currentTime - lastIncreaseTime >= increaseInterval) {
                moveAmount += 0.6f;
                lastIncreaseTime = currentTime;
            }
        }

        lastDirection = currentDirection;

        drawer.move(moveX, moveY);

    }

}

File: src/main/kotlin/actions/MovementHandler.kt
package actions

import processing.core.PApplet
import ux.PatternDrawer

// this class exists merely to encapsulate movement handling as it is a bit complex
// and i don't want to clutter the main class anymore than it already
class MovementHandler(private val drawer: PatternDrawer) {

    private var lastDirection = 0
    private val initialMoveAmount = 5f
    private var moveAmount = initialMoveAmount

    fun handleRequestedMovement() {
        for (key in pressedKeys) {
            when (key) {
                PApplet.LEFT, PApplet.RIGHT, PApplet.UP, PApplet.DOWN -> handleMovementKeys()
                else ->  // Ignore other keys and set lastDirection to 0
                    lastDirection = 0
            }
        }
    }

    private fun handleMovementKeys() {
        var moveX = 0f
        var moveY = 0f

        directions.forEach { direction ->
            if (pressedKeys.contains(direction[0])) {
                moveX += direction[2] * moveAmount / pressedKeys.size
                moveY += direction[1] * moveAmount / pressedKeys.size
            }
        }

        val currentDirection = directions.filter { pressedKeys.contains(it[0]) }.sumOf { it[0] }

        if (currentDirection != lastDirection) {
            moveAmount = initialMoveAmount
        }
        lastDirection = currentDirection

        drawer.move(moveX, moveY)
    }


    companion object {
        const val WEST = PApplet.LEFT
        const val EAST = PApplet.RIGHT
        const val NORTH = PApplet.UP
        const val SOUTH = PApplet.DOWN
        const val NORTHWEST = NORTH + WEST
        const val NORTHEAST = NORTH + EAST
        const val SOUTHWEST = SOUTH + WEST
        const val SOUTHEAST = SOUTH + EAST

        private val pressedKeys = KeyHandler.getPressedKeys()
        private val directions = arrayOf(
                intArrayOf(WEST, 0, -1),
                intArrayOf(NORTH, -1, 0),
                intArrayOf(EAST, 0, 1),
                intArrayOf(SOUTH, 1, 0),
                intArrayOf(NORTHWEST, -1, -1),
                intArrayOf(NORTHEAST, -1, 1),
                intArrayOf(SOUTHWEST, 1, -1),
                intArrayOf(SOUTHEAST, 1, 1)
        )
    }
}
